services/brig/src/Brig/API/OAuth.hs

```diff
@@ -0,0 +1,397 @@
+  rts <- lift $ wrapClient $ lookupOAuthRefreshTokens uid
+  for_ rts $ \rt -> when (oriClientId rt == cid) $ lift $ wrapClient $ deleteOAuthRefreshToken rt
+
+--------------------------------------------------------------------------------
+-- DB
+
+insertOAuthClient :: (MonadClient m) => OAuthClientId -> OAuthApplicationName -> RedirectUrl -> Password -> m ()
+insertOAuthClient cid name uri pw = retry x5 . write q $ params LocalQuorum (cid, name, uri, pw)
+  where
+    q :: PrepQuery W (OAuthClientId, OAuthApplicationName, RedirectUrl, Password) ()
+    q = "INSERT INTO oauth_client (id, name, redirect_uri, secret) VALUES (?, ?, ?, ?)"
+
+lookupOauthClient :: (MonadClient m) => OAuthClientId -> m (Maybe OAuthClient)
+lookupOauthClient cid = do
+  mNameUrl <- retry x5 . query1 q $ params LocalQuorum (Identity cid)
+  pure $ mNameUrl <&> uncurry (OAuthClient cid)
+  where
+    q :: PrepQuery R (Identity OAuthClientId) (OAuthApplicationName, RedirectUrl)
+    q = "SELECT name, redirect_uri FROM oauth_client WHERE id = ?"
+
+lookupOAuthClientSecret :: (MonadClient m) => OAuthClientId -> m (Maybe Password)
+lookupOAuthClientSecret cid = do
+  runIdentity <$$> retry x5 (query1 q (params LocalQuorum (Identity cid)))
+  where
+    q :: PrepQuery R (Identity OAuthClientId) (Identity Password)
+    q = "SELECT secret FROM oauth_client WHERE id = ?"
+
+insertOAuthAuthorizationCode :: (MonadClient m) => Word64 -> OAuthAuthorizationCode -> OAuthClientId -> UserId -> OAuthScopes -> RedirectUrl -> m ()
+insertOAuthAuthorizationCode ttl code cid uid scope uri = do
+  let cqlScope = C.Set (Set.toList (unOAuthScopes scope))
+  retry x5 . write q $ params LocalQuorum (code, cid, uid, cqlScope, uri)
+  where
+    q :: PrepQuery W (OAuthAuthorizationCode, OAuthClientId, UserId, C.Set OAuthScope, RedirectUrl) ()
+    q = fromString $ "INSERT INTO oauth_auth_code (code, client, user, scope, redirect_uri) VALUES (?, ?, ?, ?, ?) USING TTL " <> show ttl
+
+lookupOAuthAuthorizationCode :: (MonadClient m) => OAuthAuthorizationCode -> m (Maybe (OAuthClientId, UserId, OAuthScopes, RedirectUrl))
+lookupOAuthAuthorizationCode code = do
+  mTuple <- retry x5 . query1 q $ params LocalQuorum (Identity code)
+  pure $ mTuple <&> \(cid, uid, C.Set scope, uri) -> (cid, uid, OAuthScopes (Set.fromList scope), uri)
+  where
+    q :: PrepQuery R (Identity OAuthAuthorizationCode) (OAuthClientId, UserId, C.Set OAuthScope, RedirectUrl)
+    q = "SELECT client, user, scope, redirect_uri FROM oauth_auth_code WHERE code = ?"
+
+deleteOAuthAuthorizationCode :: (MonadClient m) => OAuthAuthorizationCode -> m ()
+deleteOAuthAuthorizationCode code = retry x5 . write q $ params LocalQuorum (Identity code)
+  where
+    q :: PrepQuery W (Identity OAuthAuthorizationCode) ()
+    q = "DELETE FROM oauth_auth_code WHERE code = ?"
+
+lookupAndDeleteOAuthAuthorizationCode :: (MonadClient m) => OAuthAuthorizationCode -> m (Maybe (OAuthClientId, UserId, OAuthScopes, RedirectUrl))
```

Comment:
would it make sense to store this code in hashed form as well?  we'd need the client id for that and restructure the tables, and it's only a short-lived secret, but i'm wondering...

Reply:
It's not possible because we have to be able to look up the code. If it is hashed with scrypt, that's not possible anymore.